Autopilot/Common/Interfaces/SPI.c

```diff
@@ -0,0 +1,232 @@
+/**
+ * @file SPI.h
+ * @author Ian Frosst
+ * @date March 13, 2017
+ * @copyright Waterloo Aerial Robotics Group 2017 \n
+ *   https://raw.githubusercontent.com/UWARG/PICpilot/master/LICENCE 
+ */
+
+#include <p33FJ256GP710A.h>
+#include "SPI.h"
+#include "../clock.h"
+
+typedef void (*func_ptr)(void); // this is a bad idea
+
+
+void null(){} // null function. safer than assigning pointer to NULL
+
+func_ptr SPI1_next = &null;
+func_ptr SPI2_next = &null;
+
+
+uint16_t SPI2_wait_len; // indicates how many bytes we're waiting to receive;
+
+byte SPI2_RXBF[32]; // buffer received bytes go in while we wait
+
+/*
+ * Use this file as a modular way to interface with a device over SPI.
+ */
+
+/*
+ * SPI uses 4 pins:
+ * SDIx: Serial Data Input              --> 1 : RF7, 2 : RG7
+ * SDOx: Serial Data Output             --> 1 : RF8, 2 : RG8
+ * SCKx: Shift Clock Input/Output       --> 1 : RF6, 2 : RG6
+ * SSx/FSYNCx: Slave select/Frame Sync  --> 1 : RB2, 2 : RG9
+ * Slave select is active-low
+ * 
+ * 
+ * Every TX byte/word we send, we'll expect corresponding RX data, of the same size.
+ * Whether we care about the received data is context-dependent.
+ * If we've just sent our first command, the RX bytes will be useless.
+ */
+
+// values from the data sheet
+const static uint8_t ppre_vals[4] = {64,16,4,1}; // primary pre-scale
+const static uint8_t spre_vals[8] = {8,7,6,5,4,3,2,1}; // secondary pre-scale
+
+void getPrescale(uint16_t clock, uint8_t* ppre, uint8_t* spre) {
+    int i, j;
+    uint16_t clock_khz = CLOCK_FREQUENCY / 1000;
```

Comment:
if clock is a value in herts, why do you convert the clock frequency into khz here?